File: src/ts/ssh-tcp/events/forwardedPortsCollection.ts
//
//  Copyright (c) Microsoft Corporation. All rights reserved.
//

import { SshChannel } from '@microsoft/dev-tunnels-ssh';
import { Emitter } from 'vscode-jsonrpc';
import { ForwardedPort } from './forwardedPort';
import { ForwardedPortChannelEventArgs, ForwardedPortEventArgs } from './forwardedPortEventArgs';

/**
 * Tracks the list of ports that are currently being forwarded between the SSH client and server,
 * along with the set of channel connections for each forwarded port.
 *
 * Ports forwarded in either direction (client->server or server->client) are tracked in separate
 * collections. Typically within a session the forwarding is done only in one direction, though
 * the protocol supports bi-directional forwarding.
 *
 * @see PortForwardingService.RemoteForwardedPorts
 * @see PortForwardingService.LocalForwardedPorts
 */
export class ForwardedPortsCollection implements ReadonlySet<ForwardedPort> {
	/**
	 * Maintains a mapping from port keys to port objects and channels for the port.
	 *
	 * The ForwardedPort string representation is used as the keys.
	 */
	private portChannelMap = new Map<string, [ForwardedPort, SshChannel[]]>();

	/** Gets the number of ports in the collection. */
	public get size(): number {
		return this.portChannelMap.size;
	}

	/** Checks whether a port is in the collection. */
	public has(port: ForwardedPort): boolean {
		return this.portChannelMap.has(port.toString());
	}

	/** Lists all the ports in the collection. */
	public *values(): IterableIterator<ForwardedPort> {
		for (const [port, channels] of this.portChannelMap.values()) {
			yield port;
		}
	}

	/** Iterates over all the ports in the collection. */
	public [Symbol.iterator](): IterableIterator<ForwardedPort> {
		return this.values();
	}

	/** Lists all the ports in the collection. */
	public *entries(): IterableIterator<[ForwardedPort, ForwardedPort]> {
		for (const [port, channels] of this.portChannelMap.values()) {
			yield [port, port];
		}
	}

	/**
	 * Lists all the ports in the collection.
	 * (In a set, the keys are the same as the values.)
	 */
	public keys(): IterableIterator<ForwardedPort> {
		return this.values();
	}

	/** Iterates over all the ports in the collection, invoking a callback function on each. */
	public forEach(
		callbackfn: (
			value: ForwardedPort,
			key: ForwardedPort,
			set: ReadonlySet<ForwardedPort>,
		) => void,
		thisArg?: any,
	): void {
		for (const [port, channels] of this.portChannelMap.values()) {
			callbackfn.apply(thisArg, [port, port, this]);
		}
	}

	public getChannels(port: ForwardedPort): SshChannel[] {
		const portAndChannels = this.portChannelMap.get(port.toString());
		if (!portAndChannels) {
			throw new Error(`Port ${port} is not in the collection.`);
		}

		return portAndChannels[1];
	}

	private readonly portAddedEmitter = new Emitter<ForwardedPortEventArgs>();
	public readonly onPortAdded = this.portAddedEmitter.event;

	private readonly portRemovedEmitter = new Emitter<ForwardedPortEventArgs>();
	public readonly onPortRemoved = this.portRemovedEmitter.event;

	private readonly portChannelAddedEmitter = new Emitter<ForwardedPortChannelEventArgs>();
	public readonly onPortChannelAdded = this.portChannelAddedEmitter.event;

	private readonly portChannelRemovedEmitter = new Emitter<ForwardedPortChannelEventArgs>();
	public readonly onPortChannelRemoved = this.portChannelRemovedEmitter.event;

	/** Finds the first port in the collection that matches a predicate. */
	public find(predicate: (port: ForwardedPort) => boolean): ForwardedPort | undefined {
		for (const port of this.values()) {
			if (predicate(port)) {
				return port;
			}
		}

		return undefined;
	}

	/* @internal */
	public addPort(port: ForwardedPort): void {
		if (this.has(port)) {
			throw new Error(`Port ${port} is already in the collection.`);
		}

		this.portChannelMap.set(port.toString(), [port, []]);
		this.portAddedEmitter.fire(new ForwardedPortEventArgs(port));
	}

	/* @internal */
	public removePort(port: ForwardedPort): void {
		if (!this.has(port)) {
			throw new Error(`Port ${port} is not in the collection.`);
		}

		this.portChannelMap.delete(port.toString());
		this.portRemovedEmitter.fire(new ForwardedPortEventArgs(port));
	}

	/* @internal */
	public addChannel(port: ForwardedPort, channel: SshChannel): void {
		const portAndChannels = this.portChannelMap.get(port.toString());
		if (!portAndChannels) {
			throw new Error(`Port ${port} is not in the collection.`);
		}

		const portChannels = portAndChannels[1];
		if (portChannels.find((c) => c.channelId === channel.channelId)) {
			throw new Error(
				`Channel ${channel.channelId} is already in the collection for port ${port}`,
			);
		}

		portChannels.push(channel);
		channel.onClosed(() => this.tryRemoveChannel(port, channel));
		this.portChannelAddedEmitter.fire(new ForwardedPortChannelEventArgs(port, channel));
	}

	private tryRemoveChannel(port: ForwardedPort, channel: SshChannel): void {
		const portAndChannels = this.portChannelMap.get(port.toString());
		if (portAndChannels) {
			const portChannels = portAndChannels[1];
			const index = portChannels.findIndex((c) => c.channelId === channel.channelId);
			if (index >= 0) {
				portChannels.splice(index, 1);
				this.portChannelRemovedEmitter.fire(new ForwardedPortChannelEventArgs(port, channel));
			}
		}
	}

	public toString() {
		return [...this].join(', ');
	}
}

File: src/cs/Ssh.Tcp/Events/ForwardedPortsCollection.cs
// Copyright (c) Microsoft. All rights reserved.

using System;
using System.Collections;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Diagnostics;

namespace Microsoft.DevTunnels.Ssh.Tcp.Events;

/// <summary>
/// Tracks the list of ports that are currently being forwarded between the SSH client and server,
/// along with the set of channel connections for each forwarded port.
/// </summary>
/// <remarks>
/// Ports forwarded in either direction (client->server or server->client) are tracked in separate
/// collections. Typically within a session the forwarding is done only in one direction, though
/// the protocol supports bi-directional forwarding.
/// </remarks>
/// <seealso cref="PortForwardingService.RemoteForwardedPorts" />
/// <seealso cref="PortForwardingService.LocalForwardedPorts" />
[DebuggerDisplay("{ToString(),nq}")]
public class ForwardedPortsCollection : IReadOnlyCollection<ForwardedPort>
{
	// Concurrent dictionary allows enumerating the collection while it is concurrently modified.
	private readonly ConcurrentDictionary<ForwardedPort, ConcurrentDictionary<uint, SshChannel>>
		portChannelMap = new ();

	internal ForwardedPortsCollection()
	{
	}

	/// <inheritdoc/>
	public int Count => this.portChannelMap.Count;

	/// <inheritdoc/>
	public IEnumerator<ForwardedPort> GetEnumerator() => this.portChannelMap.Keys.GetEnumerator();

	/// <inheritdoc/>
	IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();

	/// <summary>
	/// Checks whether a port is in the collection.
	/// </summary>
	public bool Contains(ForwardedPort port)
	{
		return this.portChannelMap.ContainsKey(port);
	}

	/// <summary>
	/// Gets current channels (connections) for a forwarded port.
	/// </summary>
	/// <exception cref="InvalidOperationException">The port is not in the collection.</exception>
	public IEnumerable<SshChannel> GetChannels(ForwardedPort port)
	{
		if (!this.portChannelMap.TryGetValue(port, out var channels))
		{
			throw new InvalidOperationException($"Port {port} is not in the collection.");
		}

		return channels.Values;
	}

	/// <summary>Event raised when a port is added to the collection.</summary>
	public event EventHandler<ForwardedPortEventArgs>? PortAdded;

	/// <summary>Event raised when a port in the collection is updated.</summary>
	public event EventHandler<ForwardedPortEventArgs>? PortUpdated;

	/// <summary>Event raised when a port is removed from the collection.</summary>
	public event EventHandler<ForwardedPortEventArgs>? PortRemoved;

	/// <summary>Event raised when a channel is added to the collection.</summary>
	public event EventHandler<ForwardedPortChannelEventArgs>? PortChannelAdded;

	/// <summary>Event raised when a channel is removed from the collection.</summary>
	public event EventHandler<ForwardedPortChannelEventArgs>? PortChannelRemoved;

	internal void AddOrUpdatePort(ForwardedPort port)
	{
		if (this.portChannelMap.TryAdd(port, new ConcurrentDictionary<uint, SshChannel>()))
		{
			PortAdded?.Invoke(this, new ForwardedPortEventArgs(port));
		}
		else
		{
			PortUpdated?.Invoke(this, new ForwardedPortEventArgs(port));
		}
	}

	internal void RemovePort(ForwardedPort port)
	{
		if (!this.portChannelMap.TryRemove(port, out _))
		{
			throw new InvalidOperationException($"Port {port} is not in the collection.");
		}

		PortRemoved?.Invoke(this, new ForwardedPortEventArgs(port));
	}

	internal void AddChannel(ForwardedPort port, SshChannel channel)
	{
		if (!this.portChannelMap.TryGetValue(port, out var channels))
		{
			throw new InvalidOperationException($"Port {port} is not in the collection.");
		}

		if (!channels.TryAdd(channel.ChannelId, channel))
		{
			throw new InvalidOperationException(
				$"Channel {channel.ChannelId} is already in the collection for port {port}.");
		}

		channel.Closed += (_, _) => TryRemoveChannel(port, channel);
		PortChannelAdded?.Invoke(this, new ForwardedPortChannelEventArgs(port, channel));
	}

	private void TryRemoveChannel(ForwardedPort port, SshChannel channel)
	{
		if (this.portChannelMap.TryGetValue(port, out var channels) &&
			channels.TryRemove(channel.ChannelId, out _))
		{
			PortChannelRemoved?.Invoke(this, new ForwardedPortChannelEventArgs(port, channel));
		}
	}

	public override string ToString()
	{
		return $"[{string.Join(", ", this)}]";
	}
}
